pennylane/_qubit_device.py

```diff
@@ -147,52 +212,28 @@ def statistics(self, observables):
                 results.append(self.probability(wires=obs.wires))
 
             elif obs.return_type is not None:
-                raise QuantumFunctionError("Unsupported return type specified for observable {}".format(obs.name))
+                raise QuantumFunctionError(
+                    "Unsupported return type specified for observable {}".format(obs.name)
+                )
 
         return results
 
-    def rotate_basis(self, obs_queue):
-        """Rotates the specified wires such that they
-        are in the eigenbasis of the provided observable.
-
-        Args:
-            observables (List[:class:`Observable`]): the observable queue
-        """
-        self._memory = False
-
-        wires = []
-
-        for observable in obs_queue:
-            if hasattr(observable, "return_type") and observable.return_type == Sample:
-                self._memory = True  # make sure to return samples
-
-            for diag_gate in observable.diagonalizing_gates():
-                self.apply(diag_gate)
-
-            for wire in observable.wires:
-                if isinstance(wire, int):
-                    wires.append(wire)
-                else:
-                    wires.extend(wire)
-
-        # Store the wires used by the observables such that
-        # an Identity is considered on the remaining wires
-        self._wires_used = wires
-        self._rotated_prob = self.probability(wires)
-
     def generate_samples(self):
-        """Generate computational basis samples based on the current state.
+        """Generate computational basis samples.
 
         If the device contains a sample return type, or the
         device is running in non-analytic mode, ``dev.shots`` number of
         computational basis samples are generated and stored within
         the :attr:`~._samples` attribute.
```

Comment:
I think this is not true, is it? With the current code, `generate_samples` will always be called, no matter what. I think that is an acceptable behaviour, because a plugin like `default.qubit` can overwrite this method and perform a check if sample generation is necessary.

Reply:
As far as I understood the code in `default.qubit`, `generate_samples` would now be called even in analytic mode.

Reply:
Oh, good point. In the previous version of `DefaultQubit`, we had the following in the logic:
```python
            if self._memory or (not self.analytic):
                self.generate_samples()
```

This got removed, _but_ the `self._memory` attribute and the docstrings remained. I guess we have two options:

1. Revert the logic so that `generate_samples()` is only called if 
   ```python
   self._memory = any(o.return_type == Sample for o in observables) or not self.analytic
   ```
   is true.

2. Leave it as is, and have all devices overwrite `generate_samples()` as needed (including having their own logic for when generate samples is called).

I lean slightly towards (1), because I don't see this changing much between plugins, but open to thoughts and opinions here?

Reply:
Sure, having option (1) wouldn't hinder any of the plugins from overwriting it anyways, I would also lean towards that one :)

Reply:
This has been changed to the following:
```python
    # generate computational basis samples
    if (not self.analytic) or circuit.is_sampled:
        self.generate_samples()
```